prql-compiler/src/parser/expr.rs

```diff
@@ -0,0 +1,233 @@
+
+        let expr_compare = binary_op_parser(expr_add, operator_compare());
+
+        let expr_coalesce = binary_op_parser(expr_compare, operator_coalesce());
+
+        binary_op_parser(expr_coalesce, operator_logical())
+    })
+}
+
+pub fn pipeline<E>(expr: E) -> impl Parser<Token, ExprKind, Error = Simple<Token>>
+where
+    E: Parser<Token, Expr, Error = Simple<Token>>,
+{
+    // expr is a param, so it can be either a normal expr() or
+    // a recursive expr called from within expr()
+
+    (new_line().or(whitespace()).repeated())
+        .ignore_then(
+            expr.padded_by(whitespace().or_not())
+                .separated_by(ctrl("|").or(new_line().repeated().at_least(1).to(())))
+                .at_least(1)
+                .map(|mut exprs| {
+                    if exprs.len() == 1 {
+                        exprs.remove(0).kind
+                    } else {
+                        ExprKind::Pipeline(Pipeline { exprs })
+                    }
+                }),
+        )
+        .then_ignore(new_line().or(whitespace()).repeated())
+        .labelled("pipeline")
+}
+
+fn binary_op_parser<'a, Term, Op>(
+    term: Term,
+    op: Op,
+) -> impl Parser<Token, Expr, Error = Simple<Token>> + 'a
+where
+    Term: Parser<Token, Expr, Error = Simple<Token>> + 'a,
+    Op: Parser<Token, BinOp, Error = Simple<Token>> + 'a,
+{
+    let term = term.map_with_span(|e, s| (e, s)).boxed();
+
+    (term.clone())
+        .then(op.padded_by(whitespace().or_not()).then(term).repeated())
+        .foldl(|left, (op, right)| {
+            let span = left.1.start..right.1.end;
+            let kind = ExprKind::Binary {
+                left: Box::new(left.0),
+                op,
+                right: Box::new(right.0),
+            };
+            (into_expr(kind, span.clone()), span)
+        })
+        .map(|(e, _)| e)
+        .boxed()
+}
+
+fn func_call<E>(expr: E) -> impl Parser<Token, Expr, Error = Simple<Token>>
+where
+    E: Parser<Token, Expr, Error = Simple<Token>> + Clone,
+{
+    let func = expr.clone();
+
+    let named_arg = ident_part()
+        .map(Some)
+        .then_ignore(ctrl(":").padded_by(whitespace().or_not()))
+        .then(expr.clone());
+
+    let assign_call = ident_part()
+        .then_ignore(ctrl("=").padded_by(whitespace().or_not()))
+        .then(expr.clone())
+        .map(|(alias, expr)| Expr {
+            alias: Some(alias),
+            ..expr
+        });
+    let positional_arg = assign_call.or(expr).map(|expr| (None, expr));
+
+    let args = whitespace()
+        .ignore_then(named_arg.or(positional_arg))
+        .repeated();
+
+    func.then(args)
+        .map(|(name, args)| {
+            if args.is_empty() {
+                return name.kind;
+            }
+
+            let mut named_args = HashMap::new();
+            let mut positional = Vec::new();
+            for (name, arg) in args {
+                if let Some(name) = name {
+                    named_args.insert(name, arg);
```

Comment:
You might want to check that two arguments aren't the same here by checking the result of `HashMap::insert`. If you want to report an error in here, then `.validate` instead of `.map` will allow you to do this.